### PR TITLE
Speed up packet construction

### PR DESCRIPTION
**Pull request by mikroskeem**
According to @electronicboy's benchmark, this is 6x faster than plain reflection.

### CODE REVIEW

**Comment by Aurelien30000** [comment]
You should implement that, i think it's a good idea.

**Comment by Janmm14** [comment]
The exception handling is wrong in this edit, it changes behaviour without reason.

1. Instead of retuening null if the constructor supplier is absent, you have to throw a BadPacketException.
2. The try-catch needs to be removed.

**Review by Janmm14 — CHANGES_REQUESTED** [review]
The exception handling is wrong in this edit, it changes behaviour without reason.

1. Instead of returning null if the constructor supplier is absent, you have to throw a BadPacketException.
2. The try-catch needs to be removed (not 100% sure).

**Comment by mikroskeem** [comment]
>Instead of returning null if the constructor supplier is absent, you have to throw a BadPacketException.

[Wouldn't throwing BadPacketException change the behaviour](https://github.com/SpigotMC/BungeeCord/blob/master/protocol/src/main/java/net/md_5/bungee/protocol/Protocol.java#L410)? BungeeCord does not throw anything if constructor is not present, so I shouldn't throw either.

File: BungeeCord-Patches/0054-Speed-up-packet-construction.patch [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+-            } catch ( ReflectiveOperationException ex )
++                return ( constructor == null ) ? null : constructor.get(); // Waterfall - speed up packet construction
++            } catch ( Exception ex ) // Waterfall - speed up packet construction
+             {
+                 throw new BadPacketException( "Could not construct packet with id " + id, ex );
+             }
+         }
+ 
+-        private void registerPacket(Class<? extends DefinedPacket> packetClass, ProtocolMapping... mappings)
++        public <P extends DefinedPacket> void registerPacket(Class<P> packetClass, java.util.function.Supplier<P> constructor, ProtocolMapping... mappings) // Waterfall - speed up packet construction
```
Comment (mikroskeem) [inline BungeeCord-Patches/0054-Speed-up-packet-construction.patch]: private -> public is a leftover from my fork, I didn't intend to change access modifiers.

File: BungeeCord-Patches/0054-Speed-up-packet-construction.patch [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+                     data.packetConstructors[mapping.packetID] = constructor;
+                 }
+-            } catch ( NoSuchMethodException ex )
++            } catch ( Exception ex ) // Waterfall - speed up packet construction
+             {
+                 throw new BadPacketException( "No NoArgsConstructor for packet class " + packetClass );
+             }
+         }
++        // Waterfall start - speed up packet construction (backwards compat)
++        public <P extends DefinedPacket> void registerPacket(Class<P> packetClass, ProtocolMapping... mappings) {
```
Comment (mikroskeem) [inline BungeeCord-Patches/0054-Speed-up-packet-construction.patch]: Ditto (should not be public)

**Comment by Janmm14** [comment]
@mikroskeem Sorry, I misread the old code, ofc a missing constructor should not throw since bungee doesn't handle all packet types.

**Comment by mikroskeem** [comment]
Bump

**Review by colinrgodsey — APPROVED** [review]
this all LGTM. might end up pulling it into Waterdog if it doesnt end up getting merged here

**Review by electronicboy — APPROVED** [review]

File: BungeeCord-Patches/0054-Speed-up-packet-construction.patch [hunk 3]
```diff
@@ -0,0 +1,347 @@
...
+                 Constructor<? extends DefinedPacket> constructor = packetClass.getDeclaredConstructor();
++                */ // Waterfall end
+ 
+                 int mappingIndex = 0;
+                 ProtocolMapping mapping = mappings[mappingIndex];
+@@ -479,11 +519,32 @@ public enum Protocol
+                     data.packetMap.put( packetClass, mapping.packetID );
+                     data.packetConstructors[mapping.packetID] = constructor;
+                 }
++            // Waterfall start - speed up packet construction
```
Comment (electronicboy) [inline BungeeCord-Patches/0054-Speed-up-packet-construction.patch:316]: Removed the try/catch, this woulda incorrectly wrapped exceptions here

Comment (mikroskeem) [reply BungeeCord-Patches/0054-Speed-up-packet-construction.patch:316]: Whoops, overlooked that detail :eyes: but thanks

**Pull request merged**